File: Difficulty: Medium/Count the number of possible triangles/count-the-number-of-possible-triangles.js
//{ Driver Code Starts
'use strict';

process.stdin.resume();
process.stdin.setEncoding('utf-8');

let inputString = '';
let currentLine = 0;

process.stdin.on('data', inputStdin => { inputString += inputStdin; });

process.stdin.on('end', _ => {
    inputString = inputString.trim().split('\n').map(string => string.trim());
    main();
});

function readLine() { return inputString[currentLine++]; }

function main() {
    let t = parseInt(readLine());
    for (let i = 0; i < t; i++) {
        let arr = readLine().split(' ').map(x => parseInt(x)); // Array input

        let obj = new Solution();
        console.log(obj.countTriangles(arr));
        console.log("~");
    }
}

// } Driver Code Ends

// User function Template for JavaScript

/**
 * @param {number[]} arr
 * @returns {number}
 */

class Solution {
    // Function to count the number of possible triangles
    countTriangles(arr) {
        let triangleCount = 0;

        // Iterate through all possible triplets
        for (let i = 0; i < arr.length - 2; i++) {
            for (let j = i + 1; j < arr.length - 1; j++) {
                for (let k = j + 1; k < arr.length; k++) {
                    // Check if the three sides form a valid triangle
                    if (arr[i] + arr[j] > arr[k] && 
                        arr[i] + arr[k] > arr[j] && 
                        arr[j] + arr[k] > arr[i]) {
                        triangleCount++;
                    }
                }
            }
        }

        return triangleCount;
    }
}

File: Difficulty: Medium/Count the number of possible triangles/count-the-number-of-possible-triangles.java
//{ Driver Code Starts
// Initial Template for Java

import java.io.*;
import java.util.*;

class GFG {
    public static void main(String args[]) throws IOException {
        // taking input using Scanner class
        Scanner sc = new Scanner(System.in);

        // taking total count of testcases
        int t = sc.nextInt();
        sc.nextLine();
        while (t > 0) {
            // taking elements count
            String temp[] = sc.nextLine().trim().split(" ");
            int n = temp.length;

            // creating an array of length n
            int arr[] = new int[n];

            // adding elements to the array
            for (int i = 0; i < n; i++) arr[i] = Integer.parseInt(temp[i]);

            // creating an object of class Solutions
            Solution ob = new Solution();

            // calling the method findNumberOfTriangles()
            // of class Solutions and printing the results
            System.out.println(ob.findNumberOfTriangles(arr));
            t--;
        }
    }
}

// } Driver Code Ends


// User function Template for Java

class Solution {
    // Function to count the number of possible triangles.
    static int findNumberOfTriangles(int arr[]) {
        int ans = 0;  // Variable to store the count of valid triangles
        int n = arr.length;  // Length of the input array
        
        // Sort the array to make sure we can check the triangle condition
        Arrays.sort(arr);
        
        // Iterate through the array to check all combinations of triplets
        for (int i = 0; i < n - 2; i++) {
            for (int j = i + 1; j < n - 1; j++) {
                for (int k = j + 1; k < n; k++) {
                    int a = arr[i];
                    int b = arr[j];
                    int c = arr[k];
                    
                    // Check if the three sides satisfy the triangle inequality
                    if (a + b > c && b + c > a && a + c > b) {
                        ans++;  // If valid triangle, increment the count
                    } else {
                        break;  // If the triangle inequality fails, stop checking further combinations
                    }
                }
            }
        }
        return ans;  // Return the total number of valid triangles
    }
}
